### PR TITLE
#7 - base32-encode H2 table names to support all possible DynamoDB names

### PR DESCRIPTION
**Pull request by RomanKisilenko**
Please review this PR.

### CODE REVIEW

**Comment by dmarkov** [comment]
Thanks, let me find someone who can review this pull request

**Comment by dmarkov** [comment]
@longtimeago it's yours, please review

File: src/main/java/com/jcabi/dynamo/mock/H2Data.java [hunk 1]
```diff
@@ -170,8 +171,9 @@ public H2Data(final File file) {
     public Iterable<Attributes> iterate(final String table,
         final Conditions conds) throws IOException {
         try {
+            final String h2Table = this.encodeTableName(table);
```
Comment (longtimeago) [inline src/main/java/com/jcabi/dynamo/mock/H2Data.java]: this is a one-time variable, please, avoid using it
see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#one-time-variables for details

File: src/main/java/com/jcabi/dynamo/mock/H2Data.java [hunk 2]
```diff
@@ -204,10 +206,12 @@ public H2Data(final File file) {
             throw new IOException(ex);
         }
     }
+
     @Override
     public void put(final String table, final Attributes attrs)
         throws IOException {
         try {
+            final String h2Table = this.encodeTableName(table);
```
Comment (longtimeago) [inline src/main/java/com/jcabi/dynamo/mock/H2Data.java]: same here

Comment (longtimeago) [inline src/main/java/com/jcabi/dynamo/mock/H2Data.java]: same here

File: src/main/java/com/jcabi/dynamo/mock/H2Data.java [hunk 3]
```diff
@@ -231,6 +235,7 @@ public void update(final String table, final Attributes keys,
         final AttributeUpdates attrs)
         throws IOException {
         try {
+            final String h2Table = this.encodeTableName(table);
```
Comment (longtimeago) [inline src/main/java/com/jcabi/dynamo/mock/H2Data.java]: and here

File: src/main/java/com/jcabi/dynamo/mock/H2Data.java [hunk 4]
```diff
@@ -271,8 +276,9 @@ public H2Data with(final String table, final String[] keys,
                 String.format("empty list of keys for %s table", table)
             );
         }
+        final String h2Table = this.encodeTableName(table);
```
Comment (longtimeago) [inline src/main/java/com/jcabi/dynamo/mock/H2Data.java]: and here

File: src/main/java/com/jcabi/dynamo/mock/H2Data.java [hunk 5]
```diff
@@ -319,4 +325,13 @@ private static String value(final AttributeValue attr) {
         return val;
     }
 
+    /**
+     * Base32-encodes table name for use with H2.
+     * @param table Table name to encode
+     * @return Base-32-encoded table name
+     */
+    private String encodeTableName(final String table) {
```
Comment (longtimeago) [inline src/main/java/com/jcabi/dynamo/mock/H2Data.java]: method can be static as it doesn't use instance variables

File: src/main/java/com/jcabi/dynamo/mock/H2Data.java [hunk 6]
```diff
@@ -319,4 +325,13 @@ private static String value(final AttributeValue attr) {
...
+        return new Base32(true, (byte) '_').encodeAsString(table.getBytes());
```
Comment (longtimeago) [inline src/main/java/com/jcabi/dynamo/mock/H2Data.java]: output of Base32 contains padding symbol ```=``` (see http://tools.ietf.org/html/rfc4648#section-6)
which is not allowed by AWS. Please, change padding symbol to underscore ```_``` after Bas32 conversion.
Don't forget to test

Comment (RomanKisilenko) [inline src/main/java/com/jcabi/dynamo/mock/H2Data.java]: The padding byte used on this like is already '_'.

Comment (longtimeago) [inline src/main/java/com/jcabi/dynamo/mock/H2Data.java]: you are right, looks like we are ok here

**Comment by RomanKisilenko** [comment]
@longtimeago Made changes based on review results, please check.

**Comment by longtimeago** [comment]
@rultor please, merge

File: src/test/java/com/jcabi/dynamo/mock/H2DataTest.java [hunk 7]
```diff
@@ -111,4 +113,19 @@ public void createsManyTables() throws Exception {
             .with("secondtable", new String[]{"secondid"}, new String[0]);
     }
 
+    /**
+     * H2Data can create tables with long names (max length of DynamoDb table
+     * name is 255 characters).
+     * @throws Exception In case test fails
+     */
+    @Test
```
Comment (yegor256) [inline src/test/java/com/jcabi/dynamo/mock/H2DataTest.java]: I think it's not enough to test the feature you introduced. If you remove the encoding of table names, this test will also pass. Try it. The issue is mainly with names that contains special chars, like dashes and dots. I would suggest to add another test method which will test table names `"a.b.c"` and `"a-b-c"`

**Comment by yegor256** [comment]
@RomanKisilenko one comment from me

**Comment by RomanKisilenko** [comment]
@yegor256 Unit test added.

**Comment by yegor256** [comment]
@rultor merge it

**Comment by dmarkov** [comment]
@longtimeago thanks, I just added **17 mins** to your account, payment `50616374`

+17 added to your rating, at the moment it is: [+1384](http://www.netbout.com/b/35024?open=rating)

**Comment by dmarkov** [comment]
@rultor deploy pls

**Pull request merged**